via/static/scripts/video_player/hooks/use-overflow-anchor.ts

```diff
@@ -0,0 +1,94 @@
+import { useEffect, useRef } from 'preact/hooks';
+import type { MutableRef } from 'preact/hooks';
+
+function rectIsEmpty(rect: DOMRect) {
+  return rect.width <= 0 || rect.height <= 0;
+}
+
+/**
+ * Return true if `rectB` is fully contained within `rectA`
+ */
+export function rectContains(rectA: DOMRect, rectB: DOMRect) {
+  if (rectIsEmpty(rectA) || rectIsEmpty(rectB)) {
+    return false;
+  }
+
+  return (
+    rectB.left >= rectA.left &&
+    rectB.right <= rectA.right &&
+    rectB.top >= rectA.top &&
+    rectB.bottom <= rectA.bottom
+  );
+}
+
+type ChildInViewport = {
+  element: HTMLLIElement | null;
+  offsetTop: number;
+};
+
+export function useOverflowAnchor(
+  scrollRef: MutableRef<HTMLElement | null>,
+  getChildren: (element: HTMLElement) => NodeListOf<HTMLElement>
+) {
+  const firstChildInViewport = useRef<ChildInViewport>({
+    element: null,
+    offsetTop: 0,
+  });
+
+  useEffect(() => {
+    // if (CSS.supports('overflow-anchor: auto')) {
+    //   return () => {};
+    // }
+
+    const scrollElement = scrollRef.current!;
+    let scrollRect = scrollElement.getBoundingClientRect();
+
+    const observer = new ResizeObserver(() => {
+      scrollRect = scrollElement.getBoundingClientRect();
```

Comment:
We recalculate the scrollElement's DOMRect only when it is resized. That way we can use a precalculated value on the scroll listener, which is run more frequently.